File: src/go/upack/metadata.go
package upack

import (
	"encoding/json"
	"fmt"
	"net/http"
	"net/url"
	"os"
	"strings"
)

type Metadata struct {
	PackageName    string
	Version        string
	SourceURL      string
	Authentication *[2]string
	FilePath       string
}

func (*Metadata) Name() string { return "metadata" }
func (*Metadata) Description() string {
	return "Displays metadata for a remote ProGet universal package."
}

func (m *Metadata) Help() string  { return defaultCommandHelp(m) }
func (m *Metadata) Usage() string { return defaultCommandUsage(m) }

func (*Metadata) PositionalArguments() []PositionalArgument {
	return []PositionalArgument{
		{
			Name:        "package",
			Description: "Package name and group, such as group/name.",
			Index:       0,
			TrySetValue: trySetStringValue("package", func(cmd Command) *string {
				return &cmd.(*Metadata).PackageName
			}),
		},
		{
			Name:        "version",
			Description: "Package version. If not specified, the latest version is retrieved.",
			Optional:    true,
			Index:       1,
			TrySetValue: trySetStringValue("version", func(cmd Command) *string {
				return &cmd.(*Metadata).Version
			}),
		},
	}
}

func (*Metadata) ExtraArguments() []ExtraArgument {
	return []ExtraArgument{
		{
			Name:        "source",
			Description: "URL of a upack API endpoint.",
			Required:    true,
			TrySetValue: trySetStringValue("source", func(cmd Command) *string {
				return &cmd.(*Metadata).SourceURL
			}),
		},
		{
			Name:        "user",
			Description: "User name and password to use for servers that require authentication. Example: username:password",
			TrySetValue: trySetBasicAuthValue("user", func(cmd Command) **[2]string {
				return &cmd.(*Metadata).Authentication
			}),
		},
		{
			Name:        "file",
			Description: "The metadata file to display relative to the .upack root; the default is upack.json.",
			TrySetValue: trySetStringValue("file", func(cmd Command) *string {
				return &cmd.(*Metadata).FilePath
			}),
		},
	}
}

func (m *Metadata) Run() int {
	filePath := m.FilePath
	if filePath == "" {
		filePath = "upack.json"
	}

	addr := strings.TrimRight(m.SourceURL, "/") + "/download-file/" + url.PathEscape(m.PackageName)
	if m.Version == "" {
		addr += "?latest&path=" + url.QueryEscape(filePath)
	} else {
		v, err := ParseUniversalPackageVersion(m.Version)
		if err != nil {
			fmt.Fprintln(os.Stderr, "Invalid UPack version number:", m.Version)
			return 1
		}
		addr += "/" + url.PathEscape(v.String()) + "?path=" + url.QueryEscape(filePath)
	}

	req, err := http.NewRequest("GET", addr, nil)
	if err != nil {
		fmt.Fprintln(os.Stderr, err)
		return 1
	}

	if m.Authentication != nil {
		req.SetBasicAuth(m.Authentication[0], m.Authentication[1])
	}

	resp, err := http.DefaultClient.Do(req)
	if err != nil {
		fmt.Fprintln(os.Stderr, err)
		return 1
	}
	defer resp.Body.Close()

	if resp.StatusCode >= 400 {
		fmt.Fprintln(os.Stderr, "Server returned error:", resp.Status)
		return 1
	}

	dec := json.NewDecoder(resp.Body)
	dec.UseNumber()
	token, err := dec.Token()
	if err != nil {
		fmt.Fprintln(os.Stderr, err)
		return 1
	}
	if token != json.Delim('{') {
		fmt.Fprintln(os.Stderr, "Expected JSON object")
	}
	for dec.More() {
		token, err = dec.Token()
		if err != nil {
			fmt.Fprintln(os.Stderr, err)
			return 1
		}
		key := token.(string)
		var value json.RawMessage
		err = dec.Decode(&value)
		if err != nil {
			fmt.Fprintln(os.Stderr, err)
			return 1
		}
		fmt.Printf("%s = %s\n", key, string(value))
	}

	return 0
}

File: src/dotnet/upack/Metadata.cs
using System;
using System.ComponentModel;
using System.IO;
using System.Net;
using System.Text;
using System.Threading;
using System.Threading.Tasks;
using Newtonsoft.Json;
using Newtonsoft.Json.Linq;

namespace Inedo.UPack.CLI
{
    [DisplayName("metadata")]
    [Description("Displays metadata for a remote ProGet universal package.")]
    public sealed class Metadata : Command
    {
        [DisplayName("package")]
        [Description("Package name and group, such as group/name.")]
        [PositionalArgument(0)]
        public string PackageName { get; set; }

        [DisplayName("version")]
        [Description("Package version. If not specified, the latest version is retrieved.")]
        [PositionalArgument(1, Optional = true)]
        public string Version { get; set; }

        [DisplayName("source")]
        [Description("URL of a upack API endpoint.")]
        [ExtraArgument(Optional = false)]
        [UseEnvironmentVariableAsDefault("UPACK_FEED")]
        public string SourceUrl { get; set; }

        [DisplayName("user")]
        [Description("User name and password to use for servers that require authentication. Example: \"«username»:«password»\" or \"api:«api-key»\"")]
        [ExtraArgument]
        [UseEnvironmentVariableAsDefault("UPACK_USER")]
        public NetworkCredential Authentication { get; set; }

        [DisplayName("file")]
        [Description("The metadata file to display relative to the .upack root; the default is upack.json.")]
        [ExtraArgument]
        public string FilePath { get; set; }

        public override async Task<int> RunAsync(CancellationToken cancellationToken)
        {
            var client = CreateClient(this.SourceUrl, this.Authentication);

            UniversalPackageId packageId;
            try
            {
                packageId = UniversalPackageId.Parse(this.PackageName);
            }
            catch (ArgumentException ex)
            {
                throw new UpackException("Invalid package ID: " + ex.Message, ex);
            }

            UniversalPackageVersion version = null;
            if (!string.IsNullOrEmpty(this.Version))
            {
                version = UniversalPackageVersion.TryParse(this.Version);
                if (version == null)
                    throw new UpackException($"Invalid UPack version number: {this.Version}");
            }

            JObject data;
            using (var stream = await client.GetPackageFileStreamAsync(packageId, version, string.IsNullOrEmpty(this.FilePath) ? "upack.json" : this.FilePath, cancellationToken))
            using (var reader = new StreamReader(stream, Encoding.UTF8, true, 4096, true))
            using (var jsonReader = new JsonTextReader(reader) { CloseInput = false })
            {
                data = await JObject.LoadAsync(jsonReader, cancellationToken);
            }

            foreach (var p in data.Properties())
            {
                Console.WriteLine($"{p.Name} = {p.Value}");
            }

            return 0;
        }
    }
}
